vm/src/builtins/type.rs

```diff
@@ -405,6 +405,35 @@ impl PyType {
             .unwrap_or_else(|| vm.ctx.new_str(self.name().deref()).into())
     }
 
+    #[pyproperty(magic, setter)]
+    fn set_qualname(&self, value: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
+        if value.is(&vm.ctx.none()) {
+            return Err(vm.new_type_error(format!(
+                "cannot delete '{}' attribute of immutable type '{}'",
+                "__qualname__",
+                self.name()
+            )));
+        };
+        if !self.slots.flags.has_feature(PyTypeFlags::HEAPTYPE) {
+            return Err(vm.new_type_error(format!(
+                "cannot set '{}' attribute of immutable type '{}'",
+                "__qualname__",
+                self.name()
+            )));
+        };
+        value.downcast_ref::<PyStr>().ok_or_else(|| {
+            vm.new_type_error(format!(
+                "can only assign string to {}.__qualname__, not '{}'",
+                self.name(),
+                value.class().name()
+            ))
+        })?;
```

Comment:
```suggestion
        if value.payload_if_subclass::<PyStr>(vm).is_none() {
            return Err(vm.new_type_error(format!(
                "can only assign string to {}.__qualname__, not '{}'",
                self.name(),
                value.class().name()
            )));
        }
```
It might be a useless temporary return that occurred in `downcast_ref`. As we need to check is subclass of `PyStr`, I think this is enough 😊

Reply:
Thank you. I'll fix it !